scylla-rust-wrapper/src/logging.rs

```diff
@@ -0,0 +1,121 @@
+
+pub struct Logger {
+    pub cb: CassLogCallback,
+    pub data: *mut c_void,
+}
+
+// The field `data` in the struct `Logger` is neither Send nor Sync.
+// It can be mutated only in the user-provided `CassLogCallback`, so it is safe
+// to implement Sync and Send manually for the `Logger` struct.
+unsafe impl Sync for Logger {}
+unsafe impl Send for Logger {}
+
+impl From<Level> for CassLogLevel {
+    fn from(level: Level) -> Self {
+        match level {
+            Level::INFO => CassLogLevel::CASS_LOG_INFO,
+            Level::DEBUG => CassLogLevel::CASS_LOG_DEBUG,
+            Level::WARN => CassLogLevel::CASS_LOG_WARN,
+            Level::ERROR => CassLogLevel::CASS_LOG_ERROR,
+            Level::TRACE => CassLogLevel::CASS_LOG_TRACE,
+        }
+    }
+}
+
+impl From<CassLogLevel> for Level {
+    fn from(log_level: CassLogLevel) -> Self {
+        match log_level {
+            CassLogLevel::CASS_LOG_ERROR => Level::ERROR,
+            CassLogLevel::CASS_LOG_WARN => Level::WARN,
+            CassLogLevel::CASS_LOG_INFO => Level::INFO,
+            CassLogLevel::CASS_LOG_DEBUG => Level::DEBUG,
+            CassLogLevel::CASS_LOG_TRACE => Level::TRACE,
+            _ => Level::INFO,
+        }
+    }
+}
+
+pub fn init_logging() {
+    let _log = (*LOG.read().unwrap()).as_ref().unwrap();
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn cass_log_set_level(log_level: CassLogLevel) {
+    let level = Level::try_from(log_level).ok().unwrap();
+
+    // Drops the `DefaultGuard` of the current tracing subscriber making it invalid.
+    // Setting LOG to None is vital, otherwise, the new tracing subscriber created in
+    // `set_tracing_subscriber_with_level` will be ignored as the previous `DefaultGuard` is
+    // still alive.
+    *LOG.write().unwrap() = None;
+    // Sets the tracing subscriber with new log level.
+    *LOG.write().unwrap() = Some(set_tracing_subscriber_with_level(level));
+
+    debug!("Log level is set to {}", level);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn cass_log_level_string(log_level: CassLogLevel) -> *const c_char {
+    let log_level_str = match log_level {
+        CassLogLevel::CASS_LOG_CRITICAL => "CRITICAL\0",
+        CassLogLevel::CASS_LOG_ERROR => "ERROR\0",
+        CassLogLevel::CASS_LOG_WARN => "WARN\0",
+        CassLogLevel::CASS_LOG_INFO => "INFO\0",
+        CassLogLevel::CASS_LOG_DEBUG => "DEBUG\0",
+        CassLogLevel::CASS_LOG_TRACE => "TRACE\0",
+        CassLogLevel::CASS_LOG_DISABLED => "DISABLED\0",
+        _ => "\0",
+    };
+
+    log_level_str.as_ptr() as *const c_char
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn cass_log_set_callback(callback: CassLogCallback, data: *mut c_void) {
+    let logger = match callback {
+        Some(log_cb) => Logger {
+            cb: Some(log_cb),
+            data,
+        },
+        None => Logger {
+            cb: Some(noop_log_callback),
+            data,
+        },
+    };
```

Comment:
Maybe something like this?
```suggestion
    let logger = Logger {
            cb: Some(callback.unwrap_or(noop_log_callback)),
            data,
        };
```

Reply:
This comment still stands.

Reply:
I have applied the suggestion in the same commit.